src/ast/ast_build_op_contexts.c

```diff
@@ -186,42 +142,106 @@ AST_UnwindContext AST_PrepareUnwindOp(const cypher_astnode_t *unwind_clause) {
 	return ctx;
 }
 
-AST_MergeContext AST_PrepareMergeOp(GraphContext *gc, const cypher_astnode_t *merge_clause,
-									QueryGraph *qg) {
-	const cypher_astnode_t *path = cypher_ast_merge_get_pattern_path(merge_clause);
-
-	uint entity_count = cypher_ast_pattern_path_nelements(path);
+void AST_PreparePathCreation(const cypher_astnode_t *path, QueryGraph *qg, rax *bound_vars,
+							 NodeCreateCtx **nodes, EdgeCreateCtx **edges) {
+	AST *ast = QueryCtx_GetAST();
+	GraphContext *gc = QueryCtx_GetGraphCtx();
 
-	NodeCreateCtx *nodes_to_merge = array_new(NodeCreateCtx, (entity_count / 2) + 1);
-	EdgeCreateCtx *edges_to_merge = array_new(EdgeCreateCtx, entity_count / 2);
+	// Add the path to the QueryGraph
+	QueryGraph_AddPath(qg, gc, path);
 
-	for(uint i = 0; i < entity_count; i ++) {
+	uint path_elem_count = cypher_ast_pattern_path_nelements(path);
+	for(uint i = 0; i < path_elem_count; i ++) {
+		/* See if current entity needs to be created:
+		 * 1. Current entity is NOT bound in a previous clause.
+		 * 2. We have yet to account for this entity. */
 		const cypher_astnode_t *elem = cypher_ast_pattern_path_get_element(path, i);
-		if(i % 2) {  // Entity is a relationship
+		const char *alias = AST_GetEntityName(ast, elem);
+
+		// Skip entities defined in previous clauses or already represented in our nodes/edges arrays.
+		int rc = raxTryInsert(bound_vars, (unsigned char *)alias, strlen(alias), NULL, NULL);
```

Comment:
to avoid `bound_vars` being a clone use `raxFind`

Reply:
The use of `raxTryInsert` is deliberate, because the path being inserted can introduce aliases but use them multiple times:
`CREATE (a:A)-[:E]->(a)`
Should only create one node.